docs: Vault integrations (#3449)
hashicorp/boundary: website/content/docs/integrations/vault/index.mdx

---
layout: docs

> page_title: Vault Integration

Reviewer: ```suggestion
page_title: Vault integration
```

description: |-

>  How to integrate Vault and Boundary

Reviewer: ```suggestion
 Ways to integrate Vault and Boundary
```

---

> # Boundary and Vault overview

Reviewer: ```suggestion
# Boundary and Vault integrations
```

Reviewer: There might be a better title for this. I don't think "overview" is very descriptive here, but my suggestion "integrations" isn't either. Since this topic lays out the potential ways to configure a Boundary + Vault integration, maybe "Boundary and Vault integration options" would be better, or "Boundary and Vault integration potential". What do you think?

I don't think we want to call it "How to integrate Boundary and Vault" because users will come here for step-by-step instructions, and this topic is more conceptual.

Reviewer 2: Agreed! A "How to...." implies a step by step guide. I think "Boundary and Vault Integration Options" is the better out of the two. I have plans to include a section on Terraform within this integrations section too

Reviewer: > Agreed! A "How to...." implies a step by step guide. I think "Boundary and Vault Integration Options" is the better out of the two. I have plans to include a section on Terraform within this integrations section too

Sounds good to me! 👍

> # Boundary and Vault integrations

Reviewer 3: ```suggestion
# Vault integration
```

The integration between Boundary and Vault aims to improve two main areas of concern for organizations:

- Security posture in relation to remote access
- Workflow efficiency

Boundary and Vault achieve these two goals by removing the requirement for the end-user to have to know about what the credentials are or have anything to do with using them to gain access to a resource.

> From a security perspective, we can use dynamic, ephemeral credentials that are only valid for the lifetime of the session, unless a specific Time to live (TTL) has been attached to the Vault token that would result in a session having a finite amount of time.

Reviewer 3: ```suggestion
From a security perspective, we can use dynamic, ephemeral credentials that are only valid for the lifetime of the session, unless a specific time to live (TTL) has been attached to the Vault token that would result in a session having a finite amount of time.
```

> The security benefits also extend past an organisation's internal team, but also caters for third-party or contractors that may need to access certain resources. It would be preferable to not expose credentials for resource access with temporary staff, whilst still ensuring that access is secure and granted in a timely manner.

Reviewer: Great catch @dgreeninger ! I agree that's a much stronger rewrite. Yeah, the repeat of `also` reads awkwardly there, and it is better to just cut unnecessary words. I updated my suggestion to reflect your rewrite. Thank you!

Ensuring access is granted in a timely manner then leads to that improvement in workflow efficiency. This is the result of end-users not having to be concerned with credentials, therefore removing a large bottleneck, in relation to gaining credentials through access requests and the time associated with such a task.

# Credentials
Vault can work with Boundary to be a credential [store](https://developer.hashicorp.com/boundary/docs/concepts/domain-model/credential-stores) and [library](https://developer.hashicorp.com/boundary/docs/concepts/domain-model/credential-libraries), which allows for credentials to be stored in Vault and used by Boundary. There are two options here for configuration:

- Generic Secrets.
- SSH Certificates.

> The Generic Secrets references a Vault Key-Value path, where static secrets are stored and can be used to broker credentials to the end-user when connecting to a target.

Reviewer 2: Just re-reading this bit Dan. Could we maybe have something like:

"Generic secrets reference a Vault key-value path where static secrets are stored, e.g. username and password. You can use generic secrets to broker credentials to the end-user when they connect to a target."

I just want to highlight that this is an example and then it leads nicely into brokered and referencing the dynamic element for db credentials with Vault

> The Generic Secrets references a Vault Key-Value path, where static secrets are stored and can be used to broker credentials to the end-user when connecting to a target.

Reviewer: Yeah, I like that addition! I would just change "e.g." to "for example", because it will make more sense to an international audience. We try to avoid Latin phrases.

> The Generic Secrets references a Vault Key-Value path, where static secrets are stored and can be used to broker credentials to the end-user when connecting to a target.

Reviewer 2: Cool! I'll note that for the future. Aside from that I think it's a good start!

The SSH Certificates takes advantage of Vault being the CA for SSH certificates by utilising the [HashiCorp Vault SSH Secrets Engine](https://developer.hashicorp.com/vault/docs/secrets/ssh) and can inject SSH certificates into the Boundary session.

It is worth detailing that users must bring their own Vault deployment to Boundary.

## Brokered credentials

This feature was added in Boundary 0.4 and facilitates Boundary as a credential broker for infrastructure targets, by binding credentials with user sessions, and surfacing those credentials during session initialization, with the help of Vault.

An example workflow and use case for this feature would be database access. Certain teams within an organisation, such as finance, need to access sensitive data residing in a database. Irrespective of the frequency that the database needs to be accessed, having long-lived credentials that could be purposely or accidentally leaked, can pose a huge risk to a company.

> If Boundary and Vault is added into this workflow, we would be mitigating against this potentially security risk. The Vault database secrets engine can be enabled, with the relevant database plugin configured, to then be able to issue dynamic and short-lived database credentials to the end-user. Organisations can set a limit on the amount of time these credentials are valid and once they have timed-out, access is revoked.

Reviewer: > Potentially needs to be potential, @Dan-Heath . Apologies I missed this

Good catch! I missed it too :).  I just updated my suggestion.

> If Boundary and Vault is added into this workflow, we would be mitigating against this potentially security risk. The Vault database secrets engine can be enabled, with the relevant database plugin configured, to then be able to issue dynamic and short-lived database credentials to the end-user. Organisations can set a limit on the amount of time these credentials are valid and once they have timed-out, access is revoked.

Reviewer 2: Sorry @Dan-Heath "If Boundary and Vault is" needs to be "If Boundary and Vault are"

![Brokered Credentials](/img/brokered-creds-ui.png)

Upon connecting to the database through Boundary, Boundary displays the newly generated credentials and displays to the end user to connect to the target.

## Injected credentials

Currently SSH certificate injection is the only type of injection possible with the integration between Boundary and Vault. 

> Vault is configured to act as the certificate authority (CA), to ensure that the users’ authenticity is determined by a certificate signed by a trusted CA. When key pairs are generated dynamically, they are signed by Vault to then be used to access the resources. There are two options for the key pair generation, depending on the type of Vault endpoint that is used within the Boundary credential library for SSH certification using Vault.

Reviewer: ```suggestion
Vault is configured to act as the certificate authority (CA), to ensure that the users’ authenticity is determined by a certificate signed by a trusted CA. When key pairs are generated dynamically, they are signed by Vault to then be used to access the resources. There are two options for the key pair generation, depending on the type of Vault endpoint that you use within the Boundary credential library for SSH certification.
```

Reviewer: I think the *using Vault* part at the end of the last sentence is implied, but please check me to make sure I didn't change the meaning here.

Reviewer 2: Agreed! Your edit still makes sense :)

> `/issue` - Vault will generate the key pair for you and also sign it.

Reviewer: ```suggestion
`/issue` - Vault generates and signs the key pair for you.
```

> `/sign` - The Boundary controllers will generate the key pair to then send to Vault to sign.

Reviewer: ```suggestion
`/sign` - The Boundary controllers generate the key pair and then send it to Vault to sign.
```

<Note>

> If Boundary is leveraging Vault for secrets management then one credential store equates to one Vault token. The number of Boundary targets sourcing credentials from the stores, the number of users connecting to the targets, the number of sessions that get created or how many Credential Libraries the Credential Store contains, has no impact on the client count in Vault.

Reviewer: ```suggestion
If Boundary uses Vault for secrets management, then one credential store equates to one Vault token. The number of Boundary targets that source credentials from the stores, the number of users connecting to the targets, the number of sessions that get created or how many credential libraries the credential store contains, all have no impact on the client count in Vault.
```

</Note>

> When an end user goes to connect to a target, which leverages Vault for dynamic SSH certificates, a new certificate is generated for every target connection. As long as the target trusts the CA, then access will be granted without the end-user having any visibility.

Reviewer: ```suggestion
When you connect to a target, which uses Vault for dynamic SSH certificates, a new certificate is generated for every target connection. As long as the target trusts the CA, then access is granted without you having any visibility into the credentials involved.
```
